Add option to find accounts by email

kanishk98 opened the pull request:
<!--
Remove the fields that are not appropriate
Please include:
-->

**What's this PR do?**

Adds option to find Zulip accounts by their associated email. 

**Any background context you want to provide?**

The functionality implemented uses the same API as [the current /find-accounts page](https://zulipchat.com/accounts/find). Relevant discussion may be found at https://github.com/zulip/zulip-electron/issues/600. 

**Screenshots?**

![image](https://user-images.githubusercontent.com/24617297/51748182-9c0f1a80-20d1-11e9-8f4f-f7c6868ef150.png)

**You have tested this PR on:**
  - [x] Windows
  - [ ] Linux/Ubuntu
  - [ ] macOS

kanishk98 committed:
Add front-end option to find accounts

This commit adds a field to find Zulip accounts by email to the
preferences section.

kanishk98 committed:
Add business logic to find accounts

This commit uses request() to find Zulip accounts associated with
certain emails and displays results on the screen.

Fix lint errors

rishig commented:
This should probably have a second field ("Server") that has zulipchat.com pre-filled so that it's clear it's only finding accounts from that one server.

priyank-p commented on app/renderer/js/pages/preference/find-accounts.js:
```diff
@@ -0,0 +1,83 @@
+	findAccounts() {
+		const emails = this.emailList.value;
+		if (emails && emails !== '') {
+			const url = 'https://zulipchat.com/accounts/find?emails=' + emails;
```
@kanishk98 the domain in the url here should not be hardcoded to `zulipchat.com` here, as @rishig said you should add a field that takes the server url and use that instead.

In case you didn't know, this work for any self-hosted zulip org, but the main benefit of this is mostly on `.zulipchat.com` so we should have that pre-filled. You should also use js to put the cursor in front of `.zulipchat.com`.

priyank-p commented on app/renderer/js/pages/preference/find-accounts.js:
```diff
@@ -0,0 +1,83 @@
+		`;
+	}
+
+	confirmation() {
```
How about de-duplicating confirmation and error function here, since they used same html code but diffrent messages into a new function that return the html with message like:

```javascript
	getMessageElement(msg) {
		return `
			<div class="sub-title">
				<div>${msg}</div>
			</div>
		`;
	}

	confirmation() {
		const confirmationMsg = 'Emails sent! You will only receive ' +
		'emails at addresses associated with Zulip organizations.'; 
	  	return this.getMessageElement(confirmationMsg);
	}
	
	error() {
		const errorMsg = 'Sorry, looks like something went wrong. ' +
		'Try again?';
		return this.getMessageElement(errorMsg);
	}
```

priyank-p commented on app/renderer/js/pages/preference/find-accounts.js:
```diff
@@ -0,0 +1,83 @@
+	init() {
+		this.$findAccounts = this.generateNodeFromTemplate(this.template());
+		this.$confirmation = this.generateNodeFromTemplate(this.confirmation());
+		this.$error = this.generateNodeFromTemplate(this.error());
```
I think you should only create the elements when needed, so you should be generating them when the needed below in the code path. Most of the time a user is not going to be using this and when they don't it's very unlikely both of the elements would be needed.

priyank-p commented on app/renderer/js/pages/preference/find-accounts.js:
```diff
@@ -0,0 +1,83 @@
+	}
+
+	init() {
+		this.$findAccounts = this.generateNodeFromTemplate(this.template());
```
Can't we just move this to html file and query it here? I think it would be cleaner.

akashnimare replied:
@kanishk98 ^^

priyank-p commented:
@kanishk98 It looks like this doesn't work, you should be doing a `POST` request with emails not a `GET`. I did that but the problem now is that the request get `403` error code because we don't include `crsftoken`. I don't know what do to do about this.

kanishk98 commented:
> @kanishk98 It looks like this doesn't work, you should be doing a `POST` request with emails not a `GET`. I did that but the problem now is that the request get `403` error code because we don't include `crsftoken`. I don't know what do to do about this.

@priyank-p same for me here, and this is after I tested this functionality by entering the URL into the browser and also by entering emails into the app itself. I'm not sure why it worked the first few times around. 

I'll dig around the main zulip repo and ask my questions there. Sorry for the trouble. :(

kanishk98 commented:
@priyank-p the simplest solution here would be to fire up a new `BrowserWindow` that loads the zulipchat.com URL, like so:

![image](https://user-images.githubusercontent.com/24617297/51903729-ed295200-23e2-11e9-9ccc-14ca096761a5.png)

As you can see, the `csrftoken` is present here. If you think this method shouldn't be used and we should stick to the original implementation, we _could_ consider downloading the contents of the page (I checked the Zulip repo and the `csrftoken` seems to be coming from an AJAX request, so not sure exactly how I would solve that, but I'd be happy to dig around if, like I said above, you're not in favour of a separate window.

priyank-p commented:
@kanishk98 Let's start a thread on [#desktop stream](https://chat.zulip.org/#narrow/stream/16-desktop) to discuss this further; I don't know the better model going forward with this.

kanishk98 commented:
> @kanishk98 Let's start a thread on [#desktop stream](https://chat.zulip.org/#narrow/stream/16-desktop) to discuss this further; I don't know the better model going forward with this.

There was already a thread on the stream, I've sent in a [text requesting help](https://chat.zulip.org/#narrow/stream/16-desktop/topic/Find.20accounts/near/691499).

kanishk98 commented:
I've followed [the discussion](https://chat.zulip.org/#narrow/stream/16-desktop/topic/Find.20accounts/near/696270) at czo and modified the PR to simply open the /find/account link in the user's default browser. Hope that solves this issue. 

P.S. Apologies for the repeated forced updates. `$ npm run travis` doesn't work on Windows and I typically rely on `$ npm run test` alone to give me results. Didn't work this time as expected.

priyank-p commented:
> $ npm run travis doesn't work on Windows 

@kanishk98 `npm run travis` is not meant to be run locally it's for Travis CI so that `npm run test-e2e` can work. Since we have  e2e tests disabled in CI temporarily you should just only need to rely on `npm test`, the linting, to pass.

kanishk98 commented:
@priyank-p Yes, I understand that, but I did not catch some newline errors with `npm run test` that turned up in Travis. That's the reason for my confusion. Is that a known problem as well?
If so, I can try looking into it.

priyank-p commented:
Hmm, that's weird I use windows, along with macos, to develop and I have never ran into such issue so you should defiently try to debug this. This most proabaly is some cross-platform issue that you will need to report to upstream `xo` or `eslint` project to get this solved.

akashnimare commented on app/renderer/js/pages/preference/find-accounts.js:
```diff
@@ -0,0 +1,40 @@
+
+	init() {
+		this.$findAccounts = this.generateNodeFromTemplate(this.template());
+		console.log(this.$findAccounts);
```
@kanishk98 can you remove this?

kanishk98 committed:
Replace in-app logic with open-in-browser button

Since the request to find accounts by email requires csrftoken on the
backend, we decided to just include a button to open the /accounts/find
link on the browser.

Merge branch 'master' into add-find-accounts
Fix Travis errors

Merge branch 'master' into add-find-accounts
Add newline at end of file

Merge branch 'master' into add-find-accounts
Fix trailing spaces

Merge branch 'master' into add-find-accounts

kanishk98 committed:
Move HTML code to connected-org-section

Also fix top margins

Add newline at end of file

kanishk98 committed:
Merge branch 'master' into add-find-accounts

kanishk98 commented:
@priyank-p @akashnimare I've made the necessary changes. Also saw that the margin of the find-accounts button was a little off the first time around, so I added some to the top. 
(Mentioning this because I squashed unnecessary commits and force-pushed, so the entire conversation wasn't visible)

akashnimare commented:
This looks good to me. @priyank-p can we improve it (or clean up the solution a bit)?

abhigyank commented:
Should we include the information somewhere that works only for zulipchat.com realms?

akashnimare commented:
I doubt if it's necessary since the find-account page is already pretty clear and has some info related to this.

rishig commented:
I thought the idea was the user enters their server, and it takes them to that server's find-account page?

akashnimare commented:
> I thought the idea was the user enters their server, and it takes them to that server's find-account page

Yeah, I missed this part. @kanishk98 can you add this also?

kanishk98 committed:
settings: add field for entering server URL

User can enter URL for orgs they want to find accounts for. Since this
is most useful for zulipchat.com accounts, the input box has logic for
suggesting the user enter server.zulipchat.com

Use http.startsWith to check prefix

settings: Prefill url field with zulipchat.com

kanishk98 commented:
Please let me know if the new changes work. I know Priyank had suggested that I move the `template()` code in `find-accounts.js` to the HTML in `connected-org-section.js`, but I felt that my way of doing it would maintain consistency (since `add-certificate` is also handled in the same way). Please let me know if the code structure is alright too.

priyank-p commented on app/renderer/js/pages/preference/find-accounts.js:
```diff
@@ -0,0 +1,71 @@
+
+	findAccounts(url) {
+		if (!url) {
+			return;
```
We should probably do something here like make the field border red or something?

priyank-p commented on app/renderer/js/pages/preference/find-accounts.js:
```diff
@@ -0,0 +1,71 @@
+			<div class="settings-card certificate-card">
+				<div class="certificate-input">
+					<div>Organization URL</div>
+					<input class="setting-input-value" value=".zulipchat.com"/>
```
Don't we just want the server input to have `zulipchat.com` instead of `.zulipchat.com`? cc @akashnimare @rishig.

rishig replied:
yeah

priyank-p reviewed (commented):
@kanishk98 This looks much better than the first version, Thank for working on this. I left two comment on this.

kanishk98 commented:
> We should probably do something here like make the field border red or something?

So I improved a little on Priyank's suggestion and just check if the field is empty before the user hits enter and make the field border red then. It works as below:

![4682c358c6ddf0b14c082f7557204a79 (1)](https://user-images.githubusercontent.com/24617297/56077461-19594a00-5dfa-11e9-850d-28659b8e5cca.gif)

I think we've handled the issue now. Let me know if there's anything else that should be done.

akashnimare commented on app/renderer/js/pages/preference/find-accounts.js:
```diff
@@ -0,0 +1,79 @@
+			url = 'https://' + url;
+		}
+		ipcRenderer.send('find-accounts', url + '/accounts/find');
+		ipcRenderer.on('find-accounts-reply', (event, result) => {
```
What's the use of this?

kanishk98 replied:
In `index.js`, we check if the app was able to open the URL in the browser. This happens in the following way:

```javascript
ipcMain.on('find-accounts', (event, url) => {
	const isBrowserOpened = electron.shell.openExternal(url);
	event.sender.send('find-accounts-reply', isBrowserOpened);
});
```
This code block is to notify the user if the URL entered was opened by the browser or not.

abhigyank replied:
`'Check the organization URL and try again.'` is not a relevant error message here. 
`isBrowserOpened` would be false only if there wasn't an application was available to open the URL. I think we should avoid this altogether.

abhigyank commented on app/renderer/js/pages/preference/find-accounts.js:
```diff
@@ -0,0 +1,79 @@
+		if (!url) {
+			return;
+		}
+		if (!url.includes(':')) {
```
Checking for `:` is not a good logic to determine the presence of prefix. Check the url starts with `http`. `:` can be included while specifying ports which is usually a case in self-hosted servers so that is not a good check.

kanishk98 replied:
Hmm, I didn't think of that. Thanks, I'll change it.

abhigyank commented on app/renderer/js/pages/preference/find-accounts.js:
```diff
@@ -0,0 +1,79 @@
+			<div class="settings-card certificate-card">
+				<div class="certificate-input">
+					<div>Organization URL</div>
+					<input class="setting-input-value" value="zulipchat.com"/>
```
Do we want this to be a value or placeholder? 
cc @rishig

kanishk98 replied:
From Priyank's message [here](https://github.com/zulip/zulip-electron/pull/638#discussion_r251202738), we wanted this to be pre-filled with .zulipchat.com as a value earlier, but we've since then decided to change this to zulipchat.com [here](https://github.com/zulip/zulip-electron/pull/638#discussion_r272813020). Either way, we want this to be a pre-filled value.

kanishk98 commented:
@abhigyank thanks for the review - I've made the changes you suggested. Please let me know if anything else needs to be done.

abhigyank commented on app/renderer/js/pages/preference/find-accounts.js:
```diff
@@ -0,0 +1,72 @@
+		if (!url.startsWith('http')) {
+			url = 'https://' + url;
+		}
+		ipcRenderer.send('find-accounts', url + '/accounts/find');
```
Any need to do this in the main process? We can `openExternal` at this line too.

kanishk98 replied:
Not really, I don't remember why I didn't do that. 
No problem, I'll make the changes right now.

abhigyank commented:
Other than https://github.com/zulip/zulip-electron/pull/638#discussion_r279800312,  LGTM.

kanishk98 commented:
Included [this change](https://github.com/zulip/zulip-electron/pull/638#discussion_r279800312) as a fixup to [this commit](https://github.com/zulip/zulip-electron/pull/638/commits/832df6fff19c35c2a971a2843e0850a9f8e4fd53).

kanishk98 committed:
settings: Redden URL field border if null

kanishk98 commented:
This PR has shifted direction significantly from its original idea. I think it's far simpler to send in a cleaner PR rather than continue working with this. 
I'll close this for now and send in another PR tonight with only the approved changes. 
@akashnimare let me know if you have any problems.
